File: atlas-perception-py/atlas_perception/deterministic.py
import math
import struct


def seeded_unit(seed: float) -> float:
    x = math.sin(seed * 12.9898 + 78.233) * 43758.5453
    return x - math.floor(x)


def seeded_range(seed: float, min_val: float, max_val: float) -> float:
    return min_val + seeded_unit(seed) * (max_val - min_val)


def hash_string(input_str: str) -> int:
    h = 0
    for ch in input_str:
        h = (h * 31 + ord(ch)) & 0xFFFFFFFF
    return h

File: atlas-kernel/utils/deterministic.ts
/** Deterministic pseudo-random value in [0, 1) from a numeric seed. */
export function seededUnit(seed: number): number {
  const x = Math.sin(seed * 12.9898 + 78.233) * 43758.5453;
  return x - Math.floor(x);
}

/** Deterministic value in [min, max). */
export function seededRange(seed: number, min: number, max: number): number {
  return min + seededUnit(seed) * (max - min);
}

/** Deterministic integer in [min, max]. */
export function seededInt(seed: number, min: number, max: number): number {
  return Math.floor(seededRange(seed, min, max + 1));
}

/** Hash a string into a stable numeric seed. */
export function hashString(input: string): number {
  let hash = 0;
  for (let i = 0; i < input.length; i++) {
    hash = (hash * 31 + input.charCodeAt(i)) >>> 0;
  }
  return hash;
}
